File: IAIso-v5.0/core/iaiso-java/iaiso-audit/src/main/java/io/iaiso/audit/FanoutSink.java
package io.iaiso.audit;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

/** A sink that broadcasts every event to a list of children. */
public final class FanoutSink implements Sink {
    private final List<Sink> sinks;

    public FanoutSink(Sink... sinks) {
        this.sinks = Collections.unmodifiableList(Arrays.asList(sinks));
    }

    public FanoutSink(List<Sink> sinks) {
        this.sinks = Collections.unmodifiableList(sinks);
    }

    @Override
    public void emit(Event event) {
        for (Sink s : sinks) {
            s.emit(event);
        }
    }

    public List<Sink> sinks() {
        return sinks;
    }
}

File: IAIso-v5.0/core/iaiso-php/src/Audit/FanoutSink.php
<?php

declare(strict_types=1);

namespace IAIso\Audit;

/** Forwards events to multiple sinks. */
final class FanoutSink implements Sink
{
    /** @var Sink[] */
    private array $sinks;

    public function __construct(Sink ...$sinks)
    {
        $this->sinks = $sinks;
    }

    public function emit(Event $event): void
    {
        foreach ($this->sinks as $sink) {
            try {
                $sink->emit($event);
            } catch (\Throwable) {
                // Sinks must not poison each other — swallow
            }
        }
    }
}
